source/Ship.cpp

```diff
@@ -1368,6 +1264,152 @@ void Ship::Move(vector<Visual> &visuals, list<shared_ptr<Flotsam>> &flotsam)
+		double length = hyperOffset.Length();
+		if(length > fleetRadius) // Condense fleet around leader
+		{
+			Point offsetFromDesired = hyperOffset-hyperOffset*(fleetRadius/length);
+			position -= offsetFromDesired*((0.05+0.05*hyperCount)/hyperSteps); // This calculation is arbitrary, and chosen to look good in-game
+		}
+	}
+	
+	//Hyper Drive Repeat
+	if(!isUsingJumpDrive)
+	{
+		if (hyperCount <= hyperSteps){ // If false, the ship is a cruising stage of hyperspace travel
+			velocity += (hyperAcceleration * hyperspaceRate) * angle.Unit();
+		}
+
+		if(velocity.Length() <= MaxVelocity() && !hyperspaceSystem)
+		{
+			double stoppingDistance = .5 * velocity.LengthSquared() / Acceleration(); //Distance ship travels while decelerating
+			double turnDistance = velocity.Length() * 150. / TurnRate(); //Distance ship travels before it can start decelerating
+			if (stoppingDistance+turnDistance <= (hyperTarget-position).Length())
+				hyperCount = 0;
+		}
+	}
+	
+	//Jump Drive Repeat
+	else
+	{
+		// Create the particle effects for the jump drive. This may create 100
+		// or more particles per ship per turn at the peak of the jump.
+		if(!forget)
+			CreateSparks(visuals, "jump drive", hyperCount * Width() * Height() * .000006);
+	}
+	
+	// Transition Segment
+	if(hyperCount >= hyperSteps && hyperspaceRate > 0)
+	{
+		Point hyperspaceDirection = (hyperspaceSystem->Position() - currentSystem->Position()).Unit(); // Used by Jump Drive
+		
+		currentSystem = hyperspaceSystem;
+		hyperspaceSystem = nullptr;
+		targetSystem = nullptr;
+		
+		// Make sure your velocity is in exactly the direction you are
+		// traveling in, so that when you decelerate there will not be a
+		// sudden shift in direction at the end.
+		velocity = velocity.Length() * angle.Unit();
+		
+		double borderRadius = currentSystem->Border();
+		
+		// Hyper Drive Transition
+		if(!isUsingJumpDrive)
+		{
+			Angle inverseAngle = Angle(-angle.Degrees());
+			Point targetRotated = inverseAngle.Rotate(hyperTarget); // Rotated so that positions relative to the star are from the ship's perspective. -x is left, +x is right, -y is ahead, +y is before
+			
+			double hyperStopDistance = (hyperSteps * hyperSteps) * .5 * hyperAcceleration;
+			double cruiseDistance = max(0.0, borderRadius-targetRotated.Y()+fleetRadius); // The furthest a ship could possibly have to be set back by to decelerate before the border
+			
+			position = (hyperTarget - (hyperStopDistance+cruiseDistance) * angle.Unit());
+			position += hyperOffset;
+			
+			Point offsetTargetRotated = inverseAngle.Rotate(hyperTarget+hyperOffset); // Like targetRotated, but taking into account the offset of the target.
+			if(abs(offsetTargetRotated.X()) < borderRadius)
+			{
+				cruiseDistance -= max(0.0, borderRadius*cos(asin(offsetTargetRotated.X()/borderRadius))-offsetTargetRotated.Y()); // How far into the border the ship has to go to reach its target
```

Comment:
Why are you using trig functions instead of vector math?

Reply:
Anytime you see a cos(arcsin()) or a sin(arcos()) there is almost always a better way to do it. In all my time doing maths there has never been a good reason to use either of those, the only time they might appear is while integrating.

Reply:
[cos(arcsin(x)) = sqrt(1-x^2)](http://www.wolframalpha.com/input/?i=cos(arcsin(x)))